hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -0,0 +1,112 @@
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE TypeApplications #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+module Torch.Data.Pipeline ( takeBatch
+                           , readBatches
+                           , readBatchesConcurrently
+                           , makeFold
+                           , makeFoldWithTransform
+                           , makeConcurrentFold
+                           , foldFromProducer
+                           , L.FoldM(..)
+                           , Dataset(..)
+                           , DatasetMock(..)
+                           , ConcurrentDataset(..)
+                           ) where
+
+import           Control.Arrow (first)
+import           Control.Concurrent.Async
+import qualified Control.Foldl as L
+import           Control.Monad
+import           Pipes
+import           Pipes.Concurrent
+import qualified Pipes.Prelude as P
+
+
+type Iter = Int
+type WorkerId = Int
+
+data DatasetMock m tensor = DatasetMock { getBatchMock :: Int -> m tensor
+                                        , numItersMock :: Iter
+                                        }
+
+class Dataset dataset batch  where
+  getBatch :: dataset -> Iter -> IO batch
```

Comment:
```suggestion
class (MonadPlus m) => Dataset m dataset batch where
  getBatch :: dataset -> Iter -> m batch
```

Reply:
This is definitely better than IO, but we should figure out what the best constraints `m` and `dataset` should be. `MonadPlus` seems reasonable for now.